library/src/main/java/com/google/android/exoplayer2/source/chunk/ChunkSampleStream.java

```diff
@@ -114,15 +114,30 @@ public long getBufferedPositionUs() {
     }
   }
 
+  public boolean isWithinLastChunk(long positionUs) {
+    if (isPendingReset() || loadingFinished) {
+      return false;
+    } else {
+      return ((positionUs >= getCurrentLoadPositionUs()) && (positionUs < getNextLoadPositionUs()));
+    }
+  }
+
   /**
    * Seeks to the specified position in microseconds.
    *
    * @param positionUs The seek position in microseconds.
    */
   public void seekToUs(long positionUs) {
+    boolean result;
+  
     lastSeekPositionUs = positionUs;
+    if (isWithinLastChunk(positionUs)) {
+      result = sampleQueue.skipToLastKeyframe(positionUs);
+    } else {
+      result = sampleQueue.skipToKeyframeBefore(positionUs);
+    }
     // If we're not pending a reset, see if we can seek within the sample queue.
-    boolean seekInsideBuffer = !isPendingReset() && sampleQueue.skipToKeyframeBefore(positionUs);
+    boolean seekInsideBuffer = (!isPendingReset()) && (result == true);
```

Comment:
I think the changes here are more complicated than they need to be. Why not just add a boolean argument to skipToKeyframeBefore that determines whether it's allowed to skip to the last keyframe if the position argument exceeds the last timestamp in the queue. The only change you'd need to make in skipToKeyframeBefore would be to also check this boolean argument as part of the check:

```
      if (timeUs > largestQueuedTimestampUs) {
        return C.POSITION_UNSET;
      }
```

Then here, you can simply do:

```
boolean seekInsideBuffer = !isPendingReset()
    && sampleQueue.skipToKeyframeBefore(positionUs, isWithinLastChunk(positionUs));
```

You could then revert most of this change, whilst still achieving the optimization. It's probably preferable to keep a single argument version of skipToKeyframeBefore that calls skipToKeyframeBefore(position, false), to avoid having to change any other calling sites.